### PR TITLE
NEW: in game button remapping screen Tanks demo

### PR DESCRIPTION
**Pull request by douglas-piconi**
@Rene-Damm one issue in this review.  I could not get the PerformInteractiveRebinding()  to apply the override no matter how I configured the rebind.  Spent hours trying to get it to work but it just would not apply the override when I enabled the map.   The only way I could get this to work was using WithRebindAddingNewBinding()  and removing the original binding.   Either I am missing something really obvious,  or there is a problem with the override becoming the active binding.

### CODE REVIEW

File: Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        uiButton.enabled = false;
+        okButtonMessageText.text = "Press button/stick for " + name;
+
+        var action = playerActionMap.actions[mapIndex];
+        var oldMappingPath = action.bindings[0].path;
+
+        ///REVIEW I could only get this to work by having the rebind add, and delete the old one.
+        /// the behavior of apply override does nto seem to work
+        // Remove the old binding
+        action.ChangeBindingWithPath(oldMappingPath).Erase();
```
Comment (jechter) [inline Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs]: So if you cancel the rebinding, does that mean, you will have lost your old binding?

**Comment by jechter** [comment]
I guess one thing which might be nice to add in a follow up PR is storing the bindings as preferences. Because this will be something everyone who does rebindings will want to do, and would likely look for pointers on what's the best way to do that.

File: Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        //remap and add a new binding
+        rebindOperation = action.PerformInteractiveRebinding().WithRebindAddingNewBinding().OnComplete(operation => ButtonRebindCompleted(uiButton)).Start();
+    }
+
+    private void ResetButtonMappingTextValues()
+    {
+        gasMappingValueText.text = playerActionMap.actions[gasMapIndex].bindings[0].path;
+        brakeMappingValueText.text = playerActionMap.actions[brakeMapIndex].bindings[0].path;
+        fireMappingValueText.text = playerActionMap.actions[fireMapIndex].bindings[0].path;
+        turretMappingValueText.text = playerActionMap.actions[turretMapIndex].bindings[0].path;
```
Comment (Rene-Damm) [inline Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs]: Using `InputControlPath.ToHumanReadableString()` should give very slightly more readable output.

File: Assets/TanksDemo/Assets/TanksInputActions.inputactions [hunk 3]
```diff
@@ -48,21 +48,20 @@
                     "processors": "",
                     "interactions": "Press(behavior=2)",
                     "bindings": []
+                },
+                {
+                    "name": "Pause",
+                    "id": "020abfe6-6083-46ec-9487-3aa2ecb330f3",
+                    "expectedControlLayout": "",
```
Comment (Rene-Damm) [inline Assets/TanksDemo/Assets/TanksInputActions.inputactions]: Strongly recommend setting action types / expectedControlLayouts on actions. Without them, the rebinding apparatus has no idea what it's looking for and just blindly filters for the most likely input.

**Comment by douglas-piconi** [comment]
@jechter @Rene-Damm   Ok, issues have been addressed.  Turns out I never needed the Erase(),  there is a bug with the second controller does not get the rebind, however the first one actually does.  In a separate PR I will address that issue along with more unit tests.      Rene,  effectivePath is an internal member, hence the change I made to have the binding ToString()  which resolves the effectivePath.

File: Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs (hunk 1 above)
Comment (douglas-piconi) [reply Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs]: removed the Erase, no longer an issue.

File: Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs (hunk 2 above)
Comment (douglas-piconi) [reply Assets/TanksDemo/Assets/Scripts/UI/ButtonRemapScreenController.cs]: fixed

File: Assets/TanksDemo/Assets/TanksInputActions.inputactions (hunk 3 above)
Comment (douglas-piconi) [reply Assets/TanksDemo/Assets/TanksInputActions.inputactions]: fixed

**Review by jechter — APPROVED** [review]

**Comment by Rene-Damm** [comment]
>Rene, effectivePath is an internal member

Doh, didn't realize. That shouldn't be the case. I'll add a TODO. I'm thinking that `path` should be what `effectivePath` is today. Would make the API more obvious I think.

**Review by Rene-Damm — APPROVED** [review]

**Pull request merged**